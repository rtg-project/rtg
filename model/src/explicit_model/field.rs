use super::sql_type;
use partial_struct::PartialStruct;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

#[derive(PartialStruct, Serialize, Deserialize, Debug, JsonSchema)]
#[serde(tag = "type", rename_all = "camelCase")]
#[partial(name = "ImplicitField")]
#[partial_attribute(derive(Serialize, Deserialize, Debug, JsonSchema))]
#[partial_attribute(serde(tag = "type", rename_all = "camelCase"))]
pub enum ExplicitField {
  #[serde(rename_all = "camelCase")]
  ScalarDatabaseColumn {
    name: String,
    nullable: bool,
    // #[serde(flatten)]
    sql_type: sql_type::Type,
    sql_column_name: String,
    graphql_field_name: String,
    graphql_type_name: String,
    graphql_order_by_asc: String,
    graphql_order_by_desc: String,
  },
}

// Tests
#[cfg(test)]
mod tests {
  use super::*;
  use similar_asserts::assert_eq;

  #[test]
  fn serialize() {
    let value = ExplicitField::ScalarDatabaseColumn {
      name: "drone".to_owned(),
      nullable: false,
      sql_type: sql_type::Type::Text,
      sql_column_name: "drone_col".to_owned(),
      graphql_field_name: "drone".to_owned(),
      graphql_type_name: "String".to_owned(),
      graphql_order_by_asc: "drone_ASC".to_owned(),
      graphql_order_by_desc: "drone_DESC".to_owned(),
    };

    match serde_json::to_string_pretty(&value) {
      Ok(string) => {
        assert_eq!(
          string,
          r#"{
  "type": "scalarDatabaseColumn",
  "name": "drone",
  "nullable": false,
  "sqlType": "text",
  "sqlColumnName": "drone_col",
  "graphqlFieldName": "drone",
  "graphqlTypeName": "String",
  "graphqlOrderByAsc": "drone_ASC",
  "graphqlOrderByDesc": "drone_DESC"
}"#
        );
      }
      Err(e) => panic!("{}", e),
    }
  }

  #[test]
  fn serialize_other() {
    let value = ExplicitField::ScalarDatabaseColumn {
      name: "drone".to_owned(),
      nullable: false,
      sql_type: sql_type::Type::Other("Yoo".to_owned()),
      sql_column_name: "drone_col".to_owned(),
      graphql_field_name: "drone".to_owned(),
      graphql_type_name: "String".to_owned(),
      graphql_order_by_asc: "drone_ASC".to_owned(),
      graphql_order_by_desc: "drone_DESC".to_owned(),
    };

    match serde_json::to_string_pretty(&value) {
      Ok(string) => {
        assert_eq!(
          string,
          r#"{
  "type": "scalarDatabaseColumn",
  "name": "drone",
  "nullable": false,
  "sqlType": {
    "other": "Yoo"
  },
  "sqlColumnName": "drone_col",
  "graphqlFieldName": "drone",
  "graphqlTypeName": "String",
  "graphqlOrderByAsc": "drone_ASC",
  "graphqlOrderByDesc": "drone_DESC"
}"#
        );
      }
      Err(e) => panic!("{}", e),
    }
  }

  #[test]
  fn deserialize() {
    let data = r#"
      {
        "name": "id",
        "nullable": false,
        "type": "scalarDatabaseColumn",
        "sqlType": "text",
        "sqlColumnName": "id",
        "graphqlFieldName": "id",
        "graphqlTypeName": "String",
        "graphqlOrderByAsc": "idAsc",
        "graphqlOrderByDesc": "idDesc"
      }
      "#;

    match serde_json::from_str(data) {
      Ok(field) => match field {
        ExplicitField::ScalarDatabaseColumn {
          name,
          nullable,
          sql_type,
          sql_column_name,
          graphql_field_name: _,
          graphql_type_name: _,
          graphql_order_by_asc: _,
          graphql_order_by_desc: _,
        } => {
          assert_eq!(name, "id");
          assert_eq!(nullable, false);
          assert_eq!(sql_type, sql_type::Type::Text);
          assert_eq!(sql_column_name, "id");
        }
      },
      Err(e) => panic!("{}", e),
    }
  }

  #[test]
  fn deserialize_other() {
    let data = r#"
      {
        "name": "id",
        "type": "scalarDatabaseColumn",
        "nullable": true,
        "sqlType": {
          "other": "Hii"
        },
        "sqlColumnName": "id",
        "graphqlFieldName": "id",
        "graphqlTypeName": "String",
        "graphqlOrderByAsc": "idAsc",
        "graphqlOrderByDesc": "idDesc"
      }
      "#;

    match serde_json::from_str(data) {
      Ok(field) => match field {
        ExplicitField::ScalarDatabaseColumn {
          name,
          nullable,
          sql_type,
          sql_column_name,
          graphql_field_name: _,
          graphql_type_name: _,
          graphql_order_by_asc: _,
          graphql_order_by_desc: _,
        } => {
          assert_eq!(name, "id");
          assert_eq!(nullable, true);
          assert_eq!(sql_type, sql_type::Type::Other("Hii".to_owned()));
          assert_eq!(sql_column_name, "id");
        }
      },
      Err(e) => panic!("{}", e),
    }
  }
}
