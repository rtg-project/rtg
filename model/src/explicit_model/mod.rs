pub mod entity;
pub mod field;
pub mod json_schema;
pub mod model;
pub mod sql_type;

// Tests
#[cfg(test)]
mod tests {

  use similar_asserts::assert_eq;
  #[test]
  fn it_works() {
    assert_eq!(2 + 2, 4);
  }

  #[test]
  fn it_works2() {
    assert_eq!(3 + 2, 5);
  }
}
