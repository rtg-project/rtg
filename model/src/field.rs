use super::sql_type;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Debug, JsonSchema)]
#[serde(tag = "type", rename_all = "camelCase")]
pub enum Field {
  #[serde(rename_all = "camelCase")]
  ScalarDatabaseColumn {
    name: String,
    #[serde(flatten)]
    sql_type: sql_type::Type,
    sql_column_name: String,
    graphql_field_name: String,
    graphql_type_name: String,
    graphql_order_by_asc: String,
    graphql_order_by_desc: String,
  },
}

// Tests
#[cfg(test)]
mod tests {
  use super::*;

  #[test]
  fn serialize() {
    let value = Field::ScalarDatabaseColumn {
      name: "drone".to_string(),
      sql_type: sql_type::Type::Text,
      sql_column_name: "drone_col".to_string(),
      graphql_field_name: "drone".to_string(),
      graphql_type_name: "String".to_string(),
      graphql_order_by_asc: "drone_ASC".to_string(),
      graphql_order_by_desc: "drone_DESC".to_string(),
    };

    match serde_json::to_string_pretty(&value) {
      Ok(string) => {
        assert_eq!(
          string,
          r#"{
  "type": "scalarDatabaseColumn",
  "name": "drone",
  "sqlType": "text",
  "sqlColumnName": "drone_col",
  "graphqlFieldName": "drone",
  "graphqlTypeName": "String",
  "graphqlOrderByAsc": "drone_ASC",
  "graphqlOrderByDesc": "drone_DESC"
}"#
        );
      }
      Err(e) => panic!("{}", e),
    }
  }

  #[test]
  fn serialize_other() {
    let value = Field::ScalarDatabaseColumn {
      name: "drone".to_string(),
      sql_type: sql_type::Type::Other {
        sql_type_name: "Yoo".to_string(),
      },
      sql_column_name: "drone_col".to_string(),
      graphql_field_name: "drone".to_string(),
      graphql_type_name: "String".to_string(),
      graphql_order_by_asc: "drone_ASC".to_string(),
      graphql_order_by_desc: "drone_DESC".to_string(),
    };

    match serde_json::to_string_pretty(&value) {
      Ok(string) => {
        assert_eq!(
          string,
          r#"{
  "type": "scalarDatabaseColumn",
  "name": "drone",
  "sqlType": "other",
  "sqlTypeName": "Yoo",
  "sqlColumnName": "drone_col",
  "graphqlFieldName": "drone",
  "graphqlTypeName": "String",
  "graphqlOrderByAsc": "drone_ASC",
  "graphqlOrderByDesc": "drone_DESC"
}"#
        );
      }
      Err(e) => panic!("{}", e),
    }
  }

  #[test]
  fn deserialize() {
    let data = r#"
      {
        "name": "id",
        "type": "scalarDatabaseColumn",
        "sqlType": "text",
        "sqlColumnName": "id",
        "graphqlFieldName": "id",
        "graphqlTypeName": "String",
        "graphqlOrderByAsc": "idAsc",
        "graphqlOrderByDesc": "idDesc"
      }
      "#;

    match serde_json::from_str(data) {
      Ok(field) => match field {
        Field::ScalarDatabaseColumn {
          name,
          sql_type,
          sql_column_name,
          graphql_field_name: _,
          graphql_type_name: _,
          graphql_order_by_asc: _,
          graphql_order_by_desc: _,
        } => {
          assert_eq!(name, "id");
          assert_eq!(sql_type, sql_type::Type::Text);
          assert_eq!(sql_column_name, "id");
        }
      },
      Err(e) => panic!("{}", e),
    }
  }

  #[test]
  fn deserialize_other() {
    let data = r#"
      {
        "name": "id",
        "type": "scalarDatabaseColumn",
        "sqlType": "other",
        "sqlTypeName": "Hii",
        "sqlColumnName": "id",
        "graphqlFieldName": "id",
        "graphqlTypeName": "String",
        "graphqlOrderByAsc": "idAsc",
        "graphqlOrderByDesc": "idDesc"
      }
      "#;

    match serde_json::from_str(data) {
      Ok(field) => match field {
        Field::ScalarDatabaseColumn {
          name,
          sql_type,
          sql_column_name,
          graphql_field_name: _,
          graphql_type_name: _,
          graphql_order_by_asc: _,
          graphql_order_by_desc: _,
        } => {
          assert_eq!(name, "id");
          assert_eq!(
            sql_type,
            sql_type::Type::Other {
              sql_type_name: "Hii".to_string(),
            }
          );
          assert_eq!(sql_column_name, "id");
        }
      },
      Err(e) => panic!("{}", e),
    }
  }
}
