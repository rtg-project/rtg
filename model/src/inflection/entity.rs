use inflector::Inflector;

pub fn inflect_name_from_sql_table_name(name: &str) -> String {
  Inflector::to_pascal_case(&Inflector::to_singular(name))
}

pub fn inflect_name_from_graphql_entity_type_name(name: &str) -> String {
  Inflector::to_pascal_case(&Inflector::to_singular(name))
}

pub fn inflect_sql_table_name_from_name(name: &str) -> String {
  Inflector::to_snake_case(&Inflector::to_plural(name))
}

pub fn inflect_graphql_entity_type_name_from_name(name: &str) -> String {
  Inflector::to_pascal_case(&Inflector::to_singular(name))
}

pub fn inflect_graphql_filter_type_name_from_name(name: &str) -> String {
  format!(
    "{}WhereInput",
    Inflector::to_pascal_case(&Inflector::to_singular(name))
  )
}

pub fn inflect_graphql_get_single_operation_name_from_name(name: &str) -> String {
  format!(
    "get{}",
    Inflector::to_pascal_case(&Inflector::to_singular(name))
  )
}

pub fn inflect_graphql_get_list_operation_name_from_name(name: &str) -> String {
  format!(
    "all{}",
    Inflector::to_pascal_case(&Inflector::to_plural(name))
  )
}

pub fn inflect_graphql_get_connection_operation_name_from_name(name: &str) -> String {
  format!(
    "all{}Connection",
    Inflector::to_pascal_case(&Inflector::to_plural(name))
  )
}
