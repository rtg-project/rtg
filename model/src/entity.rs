use super::field::Field;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use std::rc::Rc;

#[derive(Serialize, Deserialize, Debug, JsonSchema)]
#[serde(tag = "type", rename_all = "camelCase")]
pub enum Entity {
  #[serde(rename_all = "camelCase")]
  DatabaseTable {
    name: String,
    sql_schema_name: String,
    sql_table_name: String,
    graphql_entity_type_name: String,
    graphql_filter_type_name: String,
    graphql_get_single_operation_name: String,
    graphql_get_list_operation_name: String,
    graphql_get_connection_operation_name: String,
    graphql_default_order_by: String,
    graphql_default_first: i16,
    graphql_default_offset: i16,
    fields: Vec<Rc<Field>>,
  },
}

// Tests
#[cfg(test)]
mod tests {
  use super::*;
  use crate::sql_type;

  #[test]
  fn serialize() {
    let value = Entity::DatabaseTable {
      name: "person".to_string(),
      sql_schema_name: "public".to_string(),
      sql_table_name: "person_table".to_string(),
      graphql_entity_type_name: "Person".to_string(),
      graphql_filter_type_name: "PersonWhereFilter".to_string(),
      graphql_get_single_operation_name: "person".to_string(),
      graphql_get_list_operation_name: "persons".to_string(),
      graphql_get_connection_operation_name: "personConnection".to_string(),
      graphql_default_order_by: "id_ASC".to_string(),
      graphql_default_first: 10,
      graphql_default_offset: 0,
      fields: vec![
        Rc::new(Field::ScalarDatabaseColumn {
          name: "id".to_string(),
          sql_type: sql_type::Type::Text,
          sql_column_name: "id_col".to_string(),
          graphql_field_name: "id".to_string(),
          graphql_type_name: "String".to_string(),
          graphql_order_by_asc: "id_ASC".to_string(),
          graphql_order_by_desc: "id_DESC".to_string(),
        }),
        Rc::new(Field::ScalarDatabaseColumn {
          name: "drone".to_string(),
          sql_type: sql_type::Type::Text,
          sql_column_name: "drone_col".to_string(),
          graphql_field_name: "drone".to_string(),
          graphql_type_name: "String".to_string(),
          graphql_order_by_asc: "drone_ASC".to_string(),
          graphql_order_by_desc: "drone_DESC".to_string(),
        }),
      ],
    };

    match serde_json::to_string_pretty(&value) {
      Ok(string) => {
        assert_eq!(
          string,
          r#"{
  "type": "databaseTable",
  "name": "person",
  "sqlSchemaName": "public",
  "sqlTableName": "person_table",
  "graphqlEntityTypeName": "Person",
  "graphqlFilterTypeName": "PersonWhereFilter",
  "graphqlGetSingleOperationName": "person",
  "graphqlGetListOperationName": "persons",
  "graphqlGetConnectionOperationName": "personConnection",
  "graphqlDefaultOrderBy": "id_ASC",
  "graphqlDefaultFirst": 10,
  "graphqlDefaultOffset": 0,
  "fields": [
    {
      "type": "scalarDatabaseColumn",
      "name": "id",
      "sqlType": "text",
      "sqlColumnName": "id_col",
      "graphqlFieldName": "id",
      "graphqlTypeName": "String",
      "graphqlOrderByAsc": "id_ASC",
      "graphqlOrderByDesc": "id_DESC"
    },
    {
      "type": "scalarDatabaseColumn",
      "name": "drone",
      "sqlType": "text",
      "sqlColumnName": "drone_col",
      "graphqlFieldName": "drone",
      "graphqlTypeName": "String",
      "graphqlOrderByAsc": "drone_ASC",
      "graphqlOrderByDesc": "drone_DESC"
    }
  ]
}"#
        );
      }
      Err(e) => panic!("{}", e),
    }
  }
}
