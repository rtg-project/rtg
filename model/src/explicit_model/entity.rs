use super::field::{ExplicitField, ImplicitField};
use partial_struct::PartialStruct;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use std::rc::Rc;

#[derive(PartialStruct, Serialize, Deserialize, Debug, JsonSchema)]
#[serde(tag = "type", rename_all = "camelCase")]
#[partial(name = "ImplicitEntity")]
#[partial_attribute(derive(Serialize, Deserialize, Debug, JsonSchema))]
#[partial_attribute(serde(tag = "type", rename_all = "camelCase"))]
pub enum ExplicitEntity {
  #[serde(rename_all = "camelCase")]
  DatabaseTable {
    name: String,
    sql_schema_name: String,
    sql_table_name: String,
    graphql_entity_type_name: String,
    graphql_filter_type_name: String,
    graphql_get_single_operation_name: String,
    graphql_get_list_operation_name: String,
    graphql_get_connection_operation_name: String,
    graphql_default_order_by: String,
    graphql_default_first: i16,
    graphql_default_offset: i16,
    #[partial(nested_type = "Vec<Rc<ImplicitField>>")]
    fields: Vec<Rc<ExplicitField>>,
  },
}

// Tests
#[cfg(test)]
mod tests {
  use super::*;
  use crate::explicit_model::sql_type;
  use similar_asserts::assert_eq;

  #[test]
  fn serialize() {
    let value = ExplicitEntity::DatabaseTable {
      name: "person".to_owned(),
      sql_schema_name: "public".to_owned(),
      sql_table_name: "person_table".to_owned(),
      graphql_entity_type_name: "Person".to_owned(),
      graphql_filter_type_name: "PersonWhereFilter".to_owned(),
      graphql_get_single_operation_name: "person".to_owned(),
      graphql_get_list_operation_name: "persons".to_owned(),
      graphql_get_connection_operation_name: "personConnection".to_owned(),
      graphql_default_order_by: "id_ASC".to_owned(),
      graphql_default_first: 10,
      graphql_default_offset: 0,
      fields: vec![
        Rc::new(ExplicitField::ScalarDatabaseColumn {
          name: "id".to_owned(),
          nullable: false,
          sql_type: sql_type::Type::Text,
          sql_column_name: "id_col".to_owned(),
          graphql_field_name: "id".to_owned(),
          graphql_type_name: "String".to_owned(),
          graphql_order_by_asc: "id_ASC".to_owned(),
          graphql_order_by_desc: "id_DESC".to_owned(),
        }),
        Rc::new(ExplicitField::ScalarDatabaseColumn {
          name: "drone".to_owned(),
          nullable: false,
          sql_type: sql_type::Type::Text,
          sql_column_name: "drone_col".to_owned(),
          graphql_field_name: "drone".to_owned(),
          graphql_type_name: "String".to_owned(),
          graphql_order_by_asc: "drone_ASC".to_owned(),
          graphql_order_by_desc: "drone_DESC".to_owned(),
        }),
      ],
    };

    match serde_json::to_string_pretty(&value) {
      Ok(string) => {
        assert_eq!(
          string,
          r#"{
  "type": "databaseTable",
  "name": "person",
  "sqlSchemaName": "public",
  "sqlTableName": "person_table",
  "graphqlEntityTypeName": "Person",
  "graphqlFilterTypeName": "PersonWhereFilter",
  "graphqlGetSingleOperationName": "person",
  "graphqlGetListOperationName": "persons",
  "graphqlGetConnectionOperationName": "personConnection",
  "graphqlDefaultOrderBy": "id_ASC",
  "graphqlDefaultFirst": 10,
  "graphqlDefaultOffset": 0,
  "fields": [
    {
      "type": "scalarDatabaseColumn",
      "name": "id",
      "nullable": false,
      "sqlType": "text",
      "sqlColumnName": "id_col",
      "graphqlFieldName": "id",
      "graphqlTypeName": "String",
      "graphqlOrderByAsc": "id_ASC",
      "graphqlOrderByDesc": "id_DESC"
    },
    {
      "type": "scalarDatabaseColumn",
      "name": "drone",
      "nullable": false,
      "sqlType": "text",
      "sqlColumnName": "drone_col",
      "graphqlFieldName": "drone",
      "graphqlTypeName": "String",
      "graphqlOrderByAsc": "drone_ASC",
      "graphqlOrderByDesc": "drone_DESC"
    }
  ]
}"#
        );
      }
      Err(e) => panic!("{}", e),
    }
  }
}
