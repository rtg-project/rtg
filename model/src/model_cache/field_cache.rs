use crate::explicit_model::field::ExplicitField;
use serde::{Deserialize, Serialize};
use std::rc::Rc;

#[derive(Serialize, Deserialize, Debug)]
#[serde(tag = "type", rename_all = "camelCase")]
pub enum FieldCache {
  #[serde(rename_all = "camelCase")]
  ScalarDatabaseColumn { field: Rc<ExplicitField> },
}

impl FieldCache {
  // Another associated function, taking two arguments:
  pub fn new(field: Rc<ExplicitField>) -> FieldCache {
    match *field {
      ExplicitField::ScalarDatabaseColumn { .. } => {
        return FieldCache::ScalarDatabaseColumn {
          field: Rc::clone(&field),
        };
      }
    };
  }
}

// Tests
#[cfg(test)]
mod tests {
  use super::*;
  use crate::explicit_model::sql_type;
  use insta::{assert_debug_snapshot, assert_json_snapshot};

  #[test]
  fn constructor() {
    let value = Rc::new(ExplicitField::ScalarDatabaseColumn {
      name: "id".to_string(),
      sql_type: sql_type::Type::Text,
      sql_column_name: "id_col".to_string(),
      graphql_field_name: "id".to_string(),
      graphql_type_name: "String".to_string(),
      graphql_order_by_asc: "id_ASC".to_string(),
      graphql_order_by_desc: "id_DESC".to_string(),
    });

    assert_debug_snapshot!(FieldCache::new(Rc::clone(&value)));
    assert_json_snapshot!(
      serde_json::to_string_pretty(&FieldCache::new(Rc::clone(&value))).unwrap()
    );
  }
}
