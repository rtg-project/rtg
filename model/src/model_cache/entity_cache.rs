use super::field_cache::FieldCache;
use crate::explicit_model::entity::ExplicitEntity;
use crate::explicit_model::field::ExplicitField;
use rustc_hash::FxHashMap;
use serde::{Deserialize, Serialize};
use std::rc::Rc;

#[derive(Serialize, Deserialize, Debug)]
#[serde(tag = "type", rename_all = "camelCase")]
pub enum EntityCache {
  #[serde(rename_all = "camelCase")]
  DatabaseTable {
    entity: Rc<ExplicitEntity>,
    fields_by_sql_column_name: FxHashMap<String, Rc<FieldCache>>,
    fields_by_graphql_field_name: FxHashMap<String, Rc<FieldCache>>,
  },
}

impl EntityCache {
  // Another associated function, taking two arguments:
  pub fn new(entity: Rc<ExplicitEntity>) -> EntityCache {
    match &*entity {
      ExplicitEntity::DatabaseTable { fields, .. } => {
        let mut fields_by_sql_column_name = FxHashMap::default();
        let mut fields_by_graphql_field_name = FxHashMap::default();
        for field in fields.iter() {
          match &**field {
            ExplicitField::ScalarDatabaseColumn {
              sql_column_name,
              graphql_field_name,
              ..
            } => {
              let field_cache = Rc::new(FieldCache::new(Rc::clone(&field)));
              fields_by_sql_column_name.insert(sql_column_name.clone(), Rc::clone(&field_cache));
              fields_by_graphql_field_name
                .insert(graphql_field_name.clone(), Rc::clone(&field_cache));
            }
          }
        }
        return EntityCache::DatabaseTable {
          entity: Rc::clone(&entity),
          fields_by_sql_column_name,
          fields_by_graphql_field_name,
        };
      }
    };
  }
}

// Tests
#[cfg(test)]
mod tests {
  use super::*;
  use crate::explicit_model::sql_type;
  use insta::{assert_debug_snapshot, assert_json_snapshot};

  #[test]
  fn constructor() {
    let value = Rc::new(ExplicitEntity::DatabaseTable {
      name: "person".to_owned(),
      sql_schema_name: "public".to_owned(),
      sql_table_name: "person_table".to_owned(),
      graphql_entity_type_name: "Person".to_owned(),
      graphql_filter_type_name: "PersonWhereFilter".to_owned(),
      graphql_get_single_operation_name: "person".to_owned(),
      graphql_get_list_operation_name: "persons".to_owned(),
      graphql_get_connection_operation_name: "personConnection".to_owned(),
      graphql_default_order_by: "id_ASC".to_owned(),
      graphql_default_first: 10,
      graphql_default_offset: 0,
      fields: vec![
        Rc::new(ExplicitField::ScalarDatabaseColumn {
          name: "id".to_owned(),
          nullable: false,
          sql_type: sql_type::Type::Text,
          sql_column_name: "id_col".to_owned(),
          graphql_enabled: true,
          graphql_field_name: "id".to_owned(),
          graphql_type_name: "String".to_owned(),
          graphql_order_by_asc: "id_ASC".to_owned(),
          graphql_order_by_desc: "id_DESC".to_owned(),
        }),
        Rc::new(ExplicitField::ScalarDatabaseColumn {
          name: "drone".to_owned(),
          nullable: false,
          sql_type: sql_type::Type::Text,
          sql_column_name: "drone_col".to_owned(),
          graphql_enabled: true,
          graphql_field_name: "drone".to_owned(),
          graphql_type_name: "String".to_owned(),
          graphql_order_by_asc: "drone_ASC".to_owned(),
          graphql_order_by_desc: "drone_DESC".to_owned(),
        }),
      ],
    });

    assert_debug_snapshot!(EntityCache::new(Rc::clone(&value)));
    assert_json_snapshot!(
      serde_json::to_string_pretty(&EntityCache::new(Rc::clone(&value))).unwrap()
    );
  }
}
