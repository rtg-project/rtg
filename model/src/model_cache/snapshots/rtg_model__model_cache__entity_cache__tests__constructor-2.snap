---
source: model/src/model_cache/entity_cache.rs
expression: "serde_json::to_string_pretty(&EntityCache::new(Rc::clone(&value))).unwrap()"

---
"{\n  \"type\": \"databaseTable\",\n  \"entity\": {\n    \"type\": \"databaseTable\",\n    \"name\": \"person\",\n    \"sqlSchemaName\": \"public\",\n    \"sqlTableName\": \"person_table\",\n    \"graphqlEntityTypeName\": \"Person\",\n    \"graphqlFilterTypeName\": \"PersonWhereFilter\",\n    \"graphqlGetSingleOperationName\": \"person\",\n    \"graphqlGetListOperationName\": \"persons\",\n    \"graphqlGetConnectionOperationName\": \"personConnection\",\n    \"graphqlDefaultOrderBy\": \"id_ASC\",\n    \"graphqlDefaultFirst\": 10,\n    \"graphqlDefaultOffset\": 0,\n    \"fields\": [\n      {\n        \"type\": \"scalarDatabaseColumn\",\n        \"name\": \"id\",\n        \"sqlType\": \"text\",\n        \"sqlColumnName\": \"id_col\",\n        \"graphqlFieldName\": \"id\",\n        \"graphqlTypeName\": \"String\",\n        \"graphqlOrderByAsc\": \"id_ASC\",\n        \"graphqlOrderByDesc\": \"id_DESC\"\n      },\n      {\n        \"type\": \"scalarDatabaseColumn\",\n        \"name\": \"drone\",\n        \"sqlType\": \"text\",\n        \"sqlColumnName\": \"drone_col\",\n        \"graphqlFieldName\": \"drone\",\n        \"graphqlTypeName\": \"String\",\n        \"graphqlOrderByAsc\": \"drone_ASC\",\n        \"graphqlOrderByDesc\": \"drone_DESC\"\n      }\n    ]\n  },\n  \"fieldsBySqlColumnName\": {\n    \"id_col\": {\n      \"type\": \"scalarDatabaseColumn\",\n      \"field\": {\n        \"type\": \"scalarDatabaseColumn\",\n        \"name\": \"id\",\n        \"sqlType\": \"text\",\n        \"sqlColumnName\": \"id_col\",\n        \"graphqlFieldName\": \"id\",\n        \"graphqlTypeName\": \"String\",\n        \"graphqlOrderByAsc\": \"id_ASC\",\n        \"graphqlOrderByDesc\": \"id_DESC\"\n      }\n    },\n    \"drone_col\": {\n      \"type\": \"scalarDatabaseColumn\",\n      \"field\": {\n        \"type\": \"scalarDatabaseColumn\",\n        \"name\": \"drone\",\n        \"sqlType\": \"text\",\n        \"sqlColumnName\": \"drone_col\",\n        \"graphqlFieldName\": \"drone\",\n        \"graphqlTypeName\": \"String\",\n        \"graphqlOrderByAsc\": \"drone_ASC\",\n        \"graphqlOrderByDesc\": \"drone_DESC\"\n      }\n    }\n  },\n  \"fieldsByGraphqlFieldName\": {\n    \"drone\": {\n      \"type\": \"scalarDatabaseColumn\",\n      \"field\": {\n        \"type\": \"scalarDatabaseColumn\",\n        \"name\": \"drone\",\n        \"sqlType\": \"text\",\n        \"sqlColumnName\": \"drone_col\",\n        \"graphqlFieldName\": \"drone\",\n        \"graphqlTypeName\": \"String\",\n        \"graphqlOrderByAsc\": \"drone_ASC\",\n        \"graphqlOrderByDesc\": \"drone_DESC\"\n      }\n    },\n    \"id\": {\n      \"type\": \"scalarDatabaseColumn\",\n      \"field\": {\n        \"type\": \"scalarDatabaseColumn\",\n        \"name\": \"id\",\n        \"sqlType\": \"text\",\n        \"sqlColumnName\": \"id_col\",\n        \"graphqlFieldName\": \"id\",\n        \"graphqlTypeName\": \"String\",\n        \"graphqlOrderByAsc\": \"id_ASC\",\n        \"graphqlOrderByDesc\": \"id_DESC\"\n      }\n    }\n  }\n}"
